### DISCUSSION TITLE
Get the whole Live project in one command

### DISCUSSION THREAD

**Discussion by soundsmagic** [leolabs/ableton-js#53] [Q&A] [answered]
Hi Leo!
First of all, thank you for this application and all the work you've put into it! After studying both the Python and TS side a bit, it's obvious you've put a lot of thought into this! I started writing a WSGI server/app from scratch with no frameworks in a MIDI Remote Script, but since I now want to implement Websockets, it would be too much work to continue that path. Moving the server out to a node app with AbletonJS taking care of the server-Live-communication is dead simple (comparatively)!

Just a question: is there a way to get the whole "song state" in one command? I plan to populate my React front end in one request, and then open a Websocket connection for further changes/updates. Granted, it wouldn't be that much work to just query Live for the tracks, scenes, clips and so on, but I thought I'd ask first so I don't write code when there's already a function for this! After looking through the TS classes a bit, I haven't found anything to indicate this though, but it's certainly possible that I'm missing something. 

/John

**Comment by leolabs** [comment, marked as answer]
Hey John,

I'm happy to hear that you like the library!

Regarding your question, there's currently no way to get the entire project's state in one command, but you should be fine just querying all properties you need in parallel.

Let me know if you have any further questions :)

**Reply by soundsmagic** [reply]
Thanks for the quick reply!

I haven't worked much with concurrent stuff, could you give me a pointer regarding querying "in parallel"?

**Reply by leolabs** [reply]
That would be something like:

```ts
const [tracks, cuePoints] = await Promise.all([
  ableton.song.get("tracks"),
  ableton.song.get("cue_points"),
]);
```

`Promise.all` will run both commands in parallel and wait for both to finish before returning their results.

**Reply by soundsmagic** [reply]
Thanks again for the example! Getting more familiar with promises is certainly on my todo-list...

**Reply by leolabs** [reply]
No worries, glad I could help 😊
